src/qibolab/result.py

```diff
@@ -1,10 +1,185 @@
+            "q[V]": self.voltage_q.flatten(),
+            "phase[rad]": self.phase.flatten(),
+        }
+        return serialized_dict
+
+    @property
+    def average(self):
+        """Perform average over i and q"""
+        average_data = np.mean(self.voltage, axis=0)
+        std_data = np.std(self.voltage, axis=0, ddof=1) / np.sqrt(self.voltage.shape[0])
+        return AveragedIntegratedResults(average_data, std_data)
+
+
+class AveragedIntegratedResults(IntegratedResults):
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.INTEGRATION and AveragingMode.CYCLIC
+    or the averages of ``IntegratedResults``
+    """
+
+    def __init__(self, data: np.ndarray, std: np.ndarray = np.array([])):
+        super().__init__(data)
+        self.std: Optional[npt.NDArray[np.float64]] = std
+
+    def __add__(self, data):
+        new_res = super().__add__(data)
+        new_res.std = np.append(self.std, data.std)
+        return new_res
+
+
+class RawWaveformResults(IntegratedResults):
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.RAW and AveragingMode.SINGLESHOT
+    may also be used to store the integration weights ?
+    """
+
+
+class AveragedRawWaveformResults(AveragedIntegratedResults):
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.RAW and AveragingMode.CYCLIC
+    or the averages of ``RawWaveformResults``
+    """
+
+
+# FIXME: If probabilities are out of range the error is displeyed weirdly
+class StateResults:
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.DISCRIMINATION and AveragingMode.SINGLESHOT
+    """
+
+    def __init__(self, data: np.ndarray):
+        self.states: npt.NDArray[np.uint32] = data
+
+    def __add__(self, data):
+        return self.__class__(np.append(self.states, data.states))
+
+    def probability(self, state=0):
+        """Returns the statistical frequency of the specified state (0 or 1)."""
+        return abs(1 - state - np.mean(self.states, axis=0))
+
+    @cached_property
+    def state_0_probability(self):
+        """Returns the 0 state statistical frequency."""
+        return self.probability(0)
+
+    @cached_property
+    def state_1_probability(self):
+        """Returns the 1 state statistical frequency."""
+        return self.probability(1)
```

Comment:
Is there an advantage in writing:
```py
self.state_0_probability
```
rather than
```py
self.probability(0)
```
?

Notice that you can cache also at the level of `probability`. Even though [an actual difference exists](https://docs.python.org/3/faq/programming.html#faq-cache-method-calls).

Reply:
If it's not an incredible pain, I would do to have slightly less code around.

Instead of `@lru_cache(maxsize=20)` you can also just write [`@cache`](https://docs.python.org/3/library/functools.html#functools.cache) (it might be useful even to limit the cache, but for the time being I'd go for the simplest option).